### PR TITLE
KBP Example

### PR DESCRIPTION
**Pull request by msushkov**

### CODE REVIEW

**Comment by dennybritz** [comment]
That's great, thanks! Seems like the CSV generation was a lot of work ;)

Just a few comments:
- All lines should be <= 100 characters. That's good style (in any language) because if they are longer you cannot seem the full line without scrolling, and you cannot diff two files. I think this only applies to the application.conf, the other files are fine.
- For the escaped quotes in the application.conf file, I think it's cleaner to use """ (three quotes) to delimit the sting.
- If the weight is the same for all factors, you don't need to put a variable there, just use "?"
- Seems like `extract_features.py` is not used. If that's right, can you remove it?
- The unique extractor seems like it's not doing anything, why don't you make a view for that instead of having an extractor?

Did you talk to Ce about the example and how it's set up? I may be wrong (I haven't done this before), but in my understanding the candidate links should have features that describe how well the candidate and entity match (edit distance, overlaps, etc). To identify candidate links you could for example do a similarity join in Postgres. In your example mentions and entities have features, I'm not sure if that makes sense. How does the example in the VM work, what are the features and factors? What does Ce say?

Related to that, what does the second factor do? Did you take that from somewhere? To me it seems like it doesn't do anything, you don't have training data for the weight, and even if you did, it would affect all the variables in the same way. Also, it seems like you don't do any supervision for training data, so you cannot learn the weight for the first factor..

File: examples/kbp/application.conf [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    # Extract possible (entity, mention) pairs based on several predicates:
+    # exact string match (lowercase); Levenshtein distance < 3; similarity score >= 0.75;
+    # if one of the strings is a subset of the other
+    find_candidates: {
+      input: """
+        SELECT * FROM mention INNER JOIN entity ON
+          (lower(mention.text_contents) = lower(entity.text_contents) OR
+          levenshtein(lower(mention.text_contents), lower(entity.text_contents)) < 3 OR
+          similarity(lower(mention.text_contents), lower(entity.text_contents)) >= 0.75) OR
+          position(lower(mention.text_contents), lower(entity.text_contents)) >= 0 OR
```
Comment (dennybritz) [inline examples/kbp/application.conf]: ERROR (don't know what)

File: examples/kbp/application.conf [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  inference.factors: {
+
+    # Unary factors for a given (e, m) variable, depending on the type of feature that was extracted for the pair (e.g. edit distance)
+    exact_string_match: {
+      input_query: """SELECT * FROM link_feature AS f INNER JOIN candidate_link AS c ON
+        (f.eid = c.eid AND f.mid = c.mid)"""
+      function: "c.is_correct = Imply()"
+      weight: "?(f.feature_type)"
```
Comment (dennybritz) [inline examples/kbp/application.conf]: rename feature_type to feature

File: examples/kbp/application.conf [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    # A given mention can only link to 1 entity (no (e, m) pair with the same mid value can have
+    # the same eid value).
+    single_mention_to_one_entity: {
+      input_query: """SELECT * FROM 
+          (SELECT * FROM link_feature AS l1, link_feature AS l2 
+          WHERE l1.mid = l2.mid AND l1.eid <> l2.eid) AS t 
+          INNER JOIN candidate_link AS c ON (t.l1.eid = c.eid AND t.l1.mid = c.mid)"""
+      function: "c.is_correct = Imply()"
+      weight: "?" # the weight is the same for all instances of this rule
+    }
```
Comment (dennybritz) [inline examples/kbp/application.conf:114]: This doesn't do anything, ask Ce.

File: examples/kbp/data/README.txt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+entity/
+	entity.sql: contains a dump for the entity() table created in prepare_data.sh. Generated after
+	processing a dump of [Freebase](https://developers.google.com/freebase/data).
+
+mention/
+	mention.csv: contains the CSV data from which we will construct the mention() table specified in 
+	prepare_data.sh. Has the document id, sentence id, text contents, and mention type.
+
+queries/
```
Comment (dennybritz) [inline examples/kbp/data/README.txt]: It's outdated.

File: examples/kbp/helper_scripts/calibrate.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#! /bin/bash
+
+cd "$(dirname $0)/../..";
+ROOT_PATH=`pwd`
+
+# run calibration script and put plot into examples/kbp/output
+python $ROOT_PATH/examples/tools/cali.py $ROOT_PATH/target/calibration/link_feature.is_correct.tsv $ROOT_PATH/output/kbp_calibration.png
```
Comment (dennybritz) [inline examples/kbp/helper_scripts/calibrate.sh]: Don't need this anymore, running calibration automatically now.

File: examples/kbp/prepare_data.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	id bigserial primary key,
+	eid text references entity(eid),
+	mid bigserial references mention(id),
+	is_correct boolean);""" $DB_NAME
+
+# the feature type for the entity-mention link
+psql -c """CREATE TABLE link_feature(
+	id bigserial primary key,
+	eid text references entity(eid),
+	mid bigserial references mention(id),
```
Comment (dennybritz) [inline examples/kbp/prepare_data.sh]: Use link_id instead of eid+mid

File: examples/kbp/run.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+cd "$(dirname $0)/../..";
+ROOT_PATH=`pwd`
+
+# prepare the database: create and load tables
+$ROOT_PATH/examples/kbp/prepare_data.sh
+
+env SBT_OPTS="-Xmx4g" sbt "run -c examples/kbp/application.conf"
+
+# produce a calibration plot
+#$ROOT_PATH/examples/kbp/helper_scripts/calibrate.sh
```
Comment (dennybritz) [inline examples/kbp/run.sh]: Don't need the calibration anymore.

File: examples/kbp/udf/load_examples.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+neg_f = open(neg_ex_file, 'r')
+for line in neg_f:
+  vals = line.split(delim)
+
+  docid = vals[0].strip()
+  sentid = vals[1].strip()
+  text_contents = vals[2].strip()
+  m_type = vals[3].strip()
+
+  key = create_key(docid, sentid, text_contents)
```
Comment (dennybritz) [inline examples/kbp/udf/load_examples.py]: Use the text as the key only.

File: examples/kbp/application.conf [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+          similarity(lower(mention.text_contents), lower(entity.text_contents)) >= 0.75) OR
+          position(lower(mention.text_contents), lower(entity.text_contents)) >= 0 OR
+          position(lower(entity.text_contents), lower(mention.text_contents)))"""
+      output_relation: "candidate_link"
+      udf: "examples/kbp/udf/find_candidates.py"
+      parallelism: 4
+    }
+
+    # load the positive and negative training examples
+    load_examples: {
```
Comment (dennybritz) [inline examples/kbp/application.conf]: The supervision code should be either in the candidate link extractor, or load into a separate table. Then execute an after script to update the candidate links table with the example values.

**Comment by dennybritz** [comment]
@msushkov I just looked through the README. It sounds good, but  I think it is really important that this read like a walkthrough, not just random paragraphs that explain features. By that I mean it should exactly lay out the steps to **build** the application:
- Download DeepDive by executing this command
- Make a new folder for the example using this command
- Download the data files using this command
- Inspect the data files like this..
- Create a databse using these commands...
- Start writing a new `application.conf` and put this in there
- And so on...

The "Example application" walkthrough is the first thing people will look at when they visit the website, it's going to be the most popular page, so we need to make sure that it walks new users through the system step by step, not assuming any previous knowledge.

File: examples/kbp/udf/load_examples.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+def create_key(docid, sentid, text_contents):
+  return "<>".join([docid, sentid, text_contents])
+  
+
+'''
+Load the positive and negative examples and store in memory.
+'''
+
+pos_ex_file = "../data/training/positive_examples.csv"
+neg_ex_file = "../data/training/positive_examples.csv"
```
Comment (zifeishan) [inline examples/kbp/udf/load_examples.py]: I mean This line. Should it be:

`neg_ex_file = "../data/training/negative_examples.csv"`?
@msushkov

**Pull request closed without merging**